src/adaptations/device-layer/EFR32/freertos_bluetooth.c

```diff
@@ -0,0 +1,399 @@
+                                                                               *Source Code.
+                                                                               *
+                                                                               ******************************************************************************/
+
+#include <stdint.h>
+#include <string.h>
+#include <em_device.h>
+
+#include "rtos_gecko.h"
+#include "gecko_configuration.h"
+#include <Weave/DeviceLayer/EFR32/freertos_bluetooth.h>
+
+#ifdef CONFIGURATION_HEADER
+#include CONFIGURATION_HEADER
+#endif // CONFIGURATION_HEADER
+
+void                              BluetoothUpdate();
+volatile struct gecko_cmd_packet *bluetooth_evt;
+SemaphoreHandle_t                 BluetoothMutex = NULL;
+
+volatile static uint32_t          command_header;
+volatile static void *            command_data;
+volatile static gecko_cmd_handler command_handler_func = NULL;
+
+// Bluetooth task
+#ifndef BLUETOOTH_STACK_SIZE
+#define BLUETOOTH_STACK_SIZE (2048)
+#endif
+static void         BluetoothTask(void *p_arg);
+static TaskHandle_t BluetoothTaskHandle = NULL;
+
+// Linklayer task
+#ifndef LINKLAYER_STACK_SIZE
+#define LINKLAYER_STACK_SIZE (2048)
+#endif
+static void         LinklayerTask(void *p_arg);
+static TaskHandle_t LinklayerTaskHandle = NULL;
+//
+#define RTOS_TICK_HZ 1024
+#define BLUETOOTH_TICK_HZ 32768
+#define BLUETOOTH_TO_RTOS_TICK (BLUETOOTH_TICK_HZ / RTOS_TICK_HZ)
+
+static volatile wakeupCallback wakeupCB = NULL;
+// Set the task to post semaphore
+void BluetoothSetWakeupCallback(wakeupCallback cb)
+{
+    wakeupCB = (volatile wakeupCallback)cb;
+}
+EventGroupHandle_t bluetooth_event_flags;
+
+errorcode_t bluetooth_start(UBaseType_t               ll_priority,
+                            UBaseType_t               stack_priority,
+                            bluetooth_stack_init_func initialize_bluetooth_stack)
+{
+    errorcode_t err;
+    bluetooth_event_flags = xEventGroupCreate();
+    configASSERT(bluetooth_event_flags);
+
+    BluetoothMutex = xSemaphoreCreateMutex();
+
+    err = initialize_bluetooth_stack();
+
+    if (err == 0)
+    {
+        // create tasks for Bluetooth host stack
+        xTaskCreate(BluetoothTask,                              /* Function that implements the task. */
+                    "Bluetooth Task",                           /* Text name for the task. */
+                    BLUETOOTH_STACK_SIZE / sizeof(StackType_t), /* Number of indexes in the xStack array. */
+                    NULL,                                       /* Parameter passed into the task. */
+                    stack_priority,                             /* Priority at which the task is created. */
+                    &BluetoothTaskHandle);                      /* Variable to hold the task's data structure. */
+
+        // create tasks for Linklayer
+        xTaskCreate(LinklayerTask,                              /* Function that implements the task. */
```

Comment:
Counting the task in BLEManagerImpl, BLE support on EFR32 appears to need 3 FreeRTOS tasks.  This consumes a fair amount of resources (e.g. 5KB of stack space across all 3 tasks).  Is there any possibility of reducing the number of tasks?

The main loop for the Link Layer task seems very similar to the "Bluetooth" task, which suggests the two might be combined.  Is the separation here because of some long-running compute code in the "Bluetooth" task?

Additionally, if communication from the "Bluetooth" task to BLEManagerImpl could be orchestrated as WeaveDeviceEvents, rather than FreeRTOS event group flags, then work done in BLEManagerImpl::bluetoothStackEventHandler() could in theory be run on the Weave task, eliminating the need for the "Bluetooth App Task".

Reply:
We would like to keep it as is in this commit but agree that we should look at reducing the number of tasks if possible.  The freertos_bluetooth.c file is based on rtos_bluetooth.c in our SDK which was designed for Micrium.  The only differences are the API calls to Micrium and FreeRTOS. We may use rtos_bluetooth.c in the SDK and add ifdefs for Micrium and FreeRTOS.

Reply:
As the use of threads mirrors that of the BLE in our SDK we would be carrying significant risk changing this without changes to BLE in our SDK. So we would prefer to commit this aspect as is

Reply:
Understood.  Obviously we have the memory to support this now, so I'm fine with leaving the design as-is.  We can revisit this later as part of an effort to scale to smaller parts.